### PR TITLE
Revert BasisRotation back to being an operator

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

#9026 made `BasisRotation` a `Subroutine` so that it would be compatible with program capture QJIT. 

But since `Subroutine`'s are just special functions, they just act like a function, so the user no longer has access to the instruction object itself like they did with `Operator`. This breaks the ability to do metaprogramming, like `adjoint(op)`.  While we raised errors about this in `adjoint` and `ctrl`, there's still a bunch of places where users expect to work with the object themselves.  Therefore the `Subroutine` abstraction has now been determined to make too much of a UI change for now.

**Description of the Change:**

Extracts the parts of #9026 that are specific to `BasisRotation` and reverts them.

**Benefits:**

**Possible Drawbacks:**

`BasisRotation` is no longer compatible with capture-qjit.

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/templates/subroutines/qchem/basis_rotation.py [hunk 1]
```diff
@@ -371,19 +284,211 @@ def BasisRotation(wires: WiresLike, unitary_matrix: TensorLike, check: bool = Fa
...
+
+        if has_jax and capture.enabled():
+
+            def shift(a):
+                PhaseShift(a, wires=wires[0])
+
+            cond(jnp.logical_not(jnp.allclose(angle, 0.0)), shift)(angle)
+        else:
+            if not math.is_abstract(angle) and not math.allclose(angle, 0.0):
+                PhaseShift(angle, wires=wires[0])
```
Comment (comp-phys-marc) [inline pennylane/templates/subroutines/qchem/basis_rotation.py]: This looks a little different from what was here before.


```
if _is_jax_jit(angle):
    PhaseShift(angle, wires[0])
else:
    cond(math.logical_not(math.allclose(angle, 0.0)), PhaseShift)(angle, wires[0])
```

File: tests/capture/test_templates.py [hunk 2]
```diff
@@ -334,17 +338,6 @@ def fn(*args):
...
-    def test_basis_embedding_is_basis_state(self):
-        """Test that basis embedding is just BasisState."""
-
-        jaxpr = jax.make_jaxpr(qml.BasisEmbedding)(np.array([1, 1, 1]), wires=(0, 1, 2))
-        assert jaxpr.eqns[0].primitive == qml.BasisState._primitive
-        assert jaxpr.eqns[0].invars[0].aval == jax.core.ShapedArray((3,), int)
-
-        jaxpr = jax.make_jaxpr(qml.BasisEmbedding)(features=np.array([1, 1, 1]), wires=(0, 1, 2))
-        assert jaxpr.eqns[0].primitive == qml.BasisState._primitive
-        assert jaxpr.eqns[0].invars[0].aval == jax.core.ShapedArray((3,), int)
```
Comment (comp-phys-marc) [inline tests/capture/test_templates.py:345]: Why are we making changes to `BasisEmbedding` tests when this PR involves `BasisRotation`?

File: tests/capture/test_templates.py [hunk 3]
```diff
@@ -288,11 +292,11 @@ def fn(*args):
 # Only add a template to the following list if you manually added a test for it to
 # TestModifiedTemplates below.
 tested_modified_templates = [
-    qml.BasisEmbedding,
```
Comment (comp-phys-marc) [inline tests/capture/test_templates.py]: Why did we remove `BasisEmbedding`?

File: tests/capture/test_templates.py [hunk 4]
```diff
@@ -61,6 +61,15 @@ def normalize_for_comparison(obj):
...
     (qml.AngleEmbedding, (jnp.array([0.3, 0.1, 0.2]),), {"rotation": "Z", "wires": [0, 2, 3]}),
+    (qml.BasisEmbedding, (jnp.array([1, 0]), [2, 3]), {}),
+    pytest.param(
+        qml.BasisEmbedding,
+        (),
+        {"features": jnp.array([1, 0]), "wires": [2, 3]},
+        marks=pytest.mark.xfail(reason="arrays should never have been in the metadata [sc-104808]"),
+    ),
+    (qml.BasisEmbedding, (6, [0, 5, 2]), {"id": "my_id"}),
+    (qml.BasisEmbedding, (jnp.array([1, 0, 1]),), {"wires": [0, 2, 3]}),
```
Comment (comp-phys-marc) [inline tests/capture/test_templates.py]: Why are we adding `BasisEmbedding`?

File: tests/capture/test_templates.py [hunk 5]
```diff
@@ -872,7 +892,7 @@ def test_bbqram(self):
         """Test the primitve bind call of BBQRAM."""
 
         kwargs = {
-            "data": ((0, 1, 0), (1, 1, 1), (1, 1, 0), (0, 0, 0)),
+            "bitstrings": ("010", "111", "110", "000"),
```
Comment (comp-phys-marc) [inline tests/capture/test_templates.py]: This doesn't look right. Why are we reverting BBQRAM changes? :/

File: tests/capture/test_templates.py [hunk 6]
```diff
@@ -906,16 +926,7 @@ def test_select_only_qram(self):
...
-                (0, 1, 0),
-                (1, 1, 1),
-                (1, 1, 0),
-                (0, 0, 0),
-                (0, 1, 0),
-                (1, 1, 1),
-                (1, 1, 0),
-                (0, 0, 0),
-            ),
+            "bitstrings": ("010", "111", "110", "000", "010", "111", "110", "000"),
```
Comment (comp-phys-marc) [inline tests/capture/test_templates.py]: This probably shouldn't be reverted.

File: tests/capture/test_templates.py [hunk 7]
```diff
@@ -950,7 +961,7 @@ def test_hybrid_qram(self):
         """Test the primitve bind call of HybridQRAM."""
 
         kwargs = {
-            "data": ((0, 1, 0), (1, 1, 1), (1, 1, 0), (0, 0, 0)),
+            "bitstrings": ("010", "111", "110", "000"),
```
Comment (comp-phys-marc) [inline tests/capture/test_templates.py]: This probably shouldn't be reverted.

File: tests/capture/test_templates.py [hunk 8]
```diff
@@ -985,7 +996,7 @@ def test_qrom(self):
         """Test the primitive bind call of QROM."""
 
         kwargs = {
-            "data": ((0,), (1,)),
+            "bitstrings": ["0", "1"],
```
Comment (comp-phys-marc) [inline tests/capture/test_templates.py]: This probably shouldn't be reverted.

File: tests/capture/test_templates.py [hunk 9]
```diff
@@ -1603,8 +1608,6 @@ def test_templates_are_modified(template):
     """Test that all templates that are not listed as unmodified in the test cases above
     actually have their _primitive_bind_call modified."""
     # Make sure the template actually is modified in its primitive binding function
-    if template == qml.templates.SubroutineOp:
-        return
```
Comment (comp-phys-marc) [inline tests/capture/test_templates.py]: Do we not want to continue to ignore `Subroutine` and `SubroutineOp`? I would say if we remove those classes entirely, then we can stop ignoring them here at that time?

File: tests/ops/functions/test_equal.py [hunk 10]
```diff
@@ -1663,11 +1662,11 @@ def test_observables_equal_but_wire_order_not(self):
     def test_mid_measure(self):
         """Test that `MidMeasure`s are equal only if their wires
         an id are equal and their `reset` attribute match."""
-        mp = qml.ops.MidMeasure(wires=Wires([0]), reset=True, meas_uid="test_id")
```
Comment (comp-phys-marc) [inline tests/ops/functions/test_equal.py]: Why are we making changes to mid measure tests?

File: tests/ops/functions/test_equal.py [hunk 11]
```diff
@@ -2040,28 +2027,18 @@ def test_control_sequence_wires_comparison(self, wires1, wires2, res):
...
-        # wire types differ. match is only used if res=False, and is adjusted if res was True
-        match = "op1 and op2 have different work wires."
-        if res and wwt1 != wwt2:
-            match = "op1 and op2 have different work wire types."
-            res = False
-
-        assert qml.equal(op1, op2) is res
-
+        op1 = Controlled(base1, control_wires=2, work_wires=wire1)
+        op2 = Controlled(base2, control_wires=2, work_wires=wire2)
```
Comment (comp-phys-marc) [inline tests/ops/functions/test_equal.py]: Why are we changing controlled work wires tests?

File: tests/ops/functions/test_equal.py [hunk 12]
```diff
@@ -3158,120 +3121,3 @@ def test_select():
     op2 = qml.Select((qml.X(0),), control=2)
     qml.assert_equal(op1, op2)
     assert qml.equal(op1, op2) is True
-
-
-# pylint: disable=unused-argument
-class TestCompareSubroutines:
```
Comment (comp-phys-marc) [inline tests/ops/functions/test_equal.py:3144]: I would not remove this unless you are also removing `Subroutine` entirely.

**Review by comp-phys-marc — CHANGES_REQUESTED** [review]
Looks pretty good overall, just not sure if some things were unintentionally included that we might not intend, or if they should happen elsewhere?

File: doc/releases/changelog-dev.md [hunk 13]
```diff
@@ -136,7 +136,7 @@
   ```
 
 The following classes have been ported over:
-- `qp.BasisRotation` [(#9026)](https://github.com/PennyLaneAI/pennylane/pull/9026)
+- **REVERTED**`qml.BasisRotation` [(#9026)](https://github.com/PennyLaneAI/pennylane/pull/9026) [(#9252)](https://github.com/PennyLaneAI/pennylane/pull/9252)
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md]:
```suggestion
- **REVERTED**`qp.BasisRotation` [(#9026)](https://github.com/PennyLaneAI/pennylane/pull/9026) [(#9252)](https://github.com/PennyLaneAI/pennylane/pull/9252)
```

File: tests/capture/test_templates.py [hunk 14]
```diff
@@ -174,7 +174,7 @@ def normalize_for_comparison(obj):
     (
         qp.MottonenStatePreparation,
         (jnp.ones(8) / jnp.sqrt(8),),
-        {"wires": [3, 2, 0]},
+        {"wires": [3, 2, 0], "id": "your_id"},
```
Comment (albi3ro) [inline tests/capture/test_templates.py]:
```suggestion
        {"wires": [3, 2, 0]},
```

File: doc/releases/changelog-dev.md [hunk 15]
```diff
@@ -895,7 +895,7 @@
   ```
 
   The following classes have been ported over:
-  - `qp.BasisRotation` [(#9026)](https://github.com/PennyLaneAI/pennylane/pull/9026)
+- **REVERTED**`qp.BasisRotation` [(#9026)](https://github.com/PennyLaneAI/pennylane/pull/9026) [(#9252)](https://github.com/PennyLaneAI/pennylane/pull/92
```
Comment (JerryChen97) [inline doc/releases/changelog-dev.md]:
```suggestion
- **REVERTED**`qp.BasisRotation` [(#9026)](https://github.com/PennyLaneAI/pennylane/pull/9026) [(#9252)](https://github.com/PennyLaneAI/pennylane/pull/9252)
```

**Review by JerryChen97 — APPROVED** [review]

File: pennylane/templates/subroutines/qchem/basis_rotation.py (hunk 1 above)
Comment (comp-phys-marc) [reply pennylane/templates/subroutines/qchem/basis_rotation.py]: Thanks for fixing this.

**Review by comp-phys-marc — APPROVED** [review]
Thanks for addressing all my comments. Looks good.

File: doc/releases/changelog-dev.md [hunk 16]
```diff
@@ -940,7 +940,7 @@
   ```
 
   The following classes have been ported over:
-  - `qp.BasisRotation` [(#9026)](https://github.com/PennyLaneAI/pennylane/pull/9026)
+- **REVERTED**`qp.BasisRotation` [(#9026)](https://github.com/PennyLaneAI/pennylane/pull/9026) [(#9252)](https://github.com/PennyLaneAI/pennylane/pull/9252)
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md]:
```suggestion
  - **REVERTED**`qp.BasisRotation` [(#9026)](https://github.com/PennyLaneAI/pennylane/pull/9026) [(#9252)](https://github.com/PennyLaneAI/pennylane/pull/9252)
```
